Backup aren't shown when the site has too many activities

cleacos opened the pull request:
#### Changes proposed in this Pull Request

On Jetpack cloud and Jetpack Wpcom, sites with a bunch of activities in a short period cannot show all the full backups or the activities in either section (Backup or Activity Log) because the API responds with a maximum of 1000 items. We cannot increase this maximum to more than 1000 because the API complains about it, so with this change when we retrieve the activities we can walk through the pages (a maximum of 1000 items per page) that the API provides to us.

**Note:** We should test and monitor this patch because it changes the way we request the activities. If we have the opportunity of checking it with a site with thousands of items (more than 10000 or 20000) would be great, so we could check if it has any performance impact in the API or Calypso side.

**More info:** 1164141197617539-as-1186890594349665
**Related issues:** 5645-gh-jpop-issues

#### Testing instructions

- Apply this patch and go to a Jetpack site with Backup subscription, one with Daily and another with Real-time
- Check that everything works as before, the Backup section, and the Activity Log section.
- Find a site with a bunch of activities (more than 1000), and check that you can see all the backups and the activities
- Bonus: start a support session with a site with this issue and check that now it works well.

**Aditional test (updated: August 28)**

- Test it in several sites:
  - A new site without activities.
  - A site with a few activities.
  - A Sites with Daily Backup, Real-time Backup, and without any Backup subscription.

cleacos committed:
Playing with the filters to get only the backup activities

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-source".

@cleacos please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/843372). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

nsakaimbo commented:
👋 The `wp-desktop-source` failure is an an intermittent out-of-memory failure that's been occurring more frequently over the last week. A patch was just merged to master that bumps the available RAM to the job. Could you please rebase on latest Calypso master? Apologies for the noise!

cleacos committed:
Get all the activities from the API, retrieving from the API all the pages

cleacos commented on client/state/activity-log/utils.js:
```diff
@@ -19,7 +19,8 @@ export const filterStateToApiQuery = ( filter ) => {
 		filter.name && { name: filter.name },
-		{ number: 1000 }
+		{ number: 1000 },
+		filter.queryPage && { page: filter.queryPage }
```
We have to use `queryPage` because `filter.page` is used for paginating the UI, but at API level we use `page`

cleacos commented on client/state/data-getters/index.js:
```diff
@@ -70,8 +70,9 @@ export const getRequestActivityLogsId = ( siteId, filter ) => {
+	const queryPage = filter && filter.queryPage ? filter.queryPage : '';
 
-	return `activity-log-${ siteId }-${ group }-${ after }-${ before }-${ on }-${ aggregate }`;
+	return `activity-log-${ siteId }-${ group }-${ after }-${ before }-${ on }-${ aggregate }-${ queryPage }`;
```
To cache the query adding the `queryPage` to the query id.

cleacos commented on client/state/data-layer/wpcom/sites/activity/from-api.js:
```diff
@@ -23,7 +23,13 @@ export const DEFAULT_GRIDICON = 'info-outline';
 export function transformer( apiResponse ) {
-	return get( apiResponse, [ 'current', 'orderedItems' ], [] ).map( processItem );
+	const response = {
+		totalItems: apiResponse.totalItems,
+		totalPages: apiResponse.totalPages,
+		page: apiResponse.page,
```
We add to the object we transform the data received to detect the total pages and the current page of the response to the request.

cleacos committed:
Pass the state of the response to the new object

cleacos committed:
If totalPages is zero stop the page requests

cleacos committed:
Fix: use totalItems to check if we should stop the requests

cleacos committed:
Revert the last commit because it is a asyc request

We don't have the data get, so we cannot check it.

cleacos committed:
After the page 5 we get the next activities with nextAfter

Note: for some reason the API stop the responses after some iterations through the result because the `search_after` value is equal to `nextAfter`, so we have to stop the loop there.

mattgawarecki commented:
I was able to test the latest version of this pull request, and it's a significant improvement over what we have. Sites with lots of activities now have at least slightly better visibility into the past, but as @cleacos notes, we have a back-end issue to attend to before a permanent fix is possible.

I'm approving this for deployment, with the idea that we'll iterate and improve toward an even better user experience.

mattgawarecki commented:
Closing in favor of #46932 :-)
